pkg/controller/master/rancher/pod_resources.go

```diff
@@ -0,0 +1,132 @@
+package rancher
+
+import (
+	"encoding/json"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+// Since multi-cluster-management-agent is disabled in the Harvester embedded Rancher,
+// copy and modified this file from Rancher to fix https://github.com/harvester/harvester/issues/2310
+// source file: https://github.com/rancher/rancher/tree/release/v2.6/pkg/controllers/managementagent/podresources
+
+const (
+	RequestsAnnotation = "management.cattle.io/pod-requests"
+	LimitsAnnotation   = "management.cattle.io/pod-limits"
+)
+
+func (h *Handler) PodResourcesOnChanged(key string, node *corev1.Node) (*corev1.Node, error) {
+	if node == nil {
+		return node, nil
+	}
+
+	h.nodeController.EnqueueAfter(node.Name, 15*time.Second)
+
+	pods, err := h.getNonTerminatedPods(node)
+	if err != nil {
+		return nil, err
+	}
+
+	requests, limits, err := getPodResourceAnnotations(pods)
+	if err != nil {
+		return nil, err
+	}
+
+	if node.Annotations[RequestsAnnotation] != requests ||
+		node.Annotations[LimitsAnnotation] != limits {
+		toUpdate := node.DeepCopy()
+		if toUpdate.Annotations == nil {
+			toUpdate.Annotations = map[string]string{}
+		}
+		toUpdate.Annotations[RequestsAnnotation] = requests
+		toUpdate.Annotations[LimitsAnnotation] = limits
+		return h.nodeController.Update(toUpdate)
+	}
+
+	return node, nil
+}
+
+func (h *Handler) getNonTerminatedPods(node *corev1.Node) ([]*corev1.Pod, error) {
+	var pods []*corev1.Pod
+	fromCache, err := h.podCache.List("", labels.NewSelector())
```

Comment:
Updated in a new commit https://github.com/harvester/harvester/pull/2483/commits/aec64208131cbbd35b8db111d48c92e2af979304

Reply:
A bit surprised that the `Rancher` code is simply retriving all PODs. Given 100 NODEs * 30 PODs, it will retrieve and loop 3000 PODs every 15 seconds.

```
rancher: pkg/controllers/managementagent/podresources/register.go
func (p *podResources) getNonTerminatedPods(node *corev1.Node) ([]*corev1.Pod, error) {
	var pods []*corev1.Pod
	fromCache, err := p.podLister.List("", labels.NewSelector())
..
	for _, pod := range fromCache {
		if pod.Spec.NodeName != node.Name {
			continue
		}	
```